src/hooks/useCartRequest.ts

```diff
@@ -0,0 +1,58 @@
+import { CartListAction } from 'redux/cartList/action';
+import {
+  deleteAllCartItemRequest,
+  deleteCartItemRequest,
+  patchAllCartSelectedRequest,
+  patchCartSelectedRequest,
+  postCartItemRequest,
+  putCartItemRequest,
+} from 'redux/cartList/thunk';
+import { CartItem } from 'types/domain';
+
+import { useAppDispatch } from './useAppDispatch';
+
+const useCartRequest = (cartList: CartItem[]) => {
+  const dispatch = useAppDispatch<CartListAction>();
+
+  const postCartItemQuantity =
+    (id: number | string) =>
+    (diff = 1) => {
+      dispatch(postCartItemRequest({ id: Number(id), quantity: diff, isSelected: true }));
+    };
+
+  const updateCartItemQuantity =
+    (id: number | string) =>
+    (diff = 1) => {
+      const targetId = typeof id === 'number' ? id : Number(id);
+      const targetItem = cartList.find(cartItem => cartItem.id === targetId);
+
+      dispatch(putCartItemRequest({ ...targetItem, quantity: targetItem.quantity + diff }));
+    };
+
+  const selectCartItem = (id: number | string) => () => {
+    dispatch(patchCartSelectedRequest(Number(id)));
+  };
+
+  const selectAllCartItem = (isAllSelected: boolean) => {
+    dispatch(patchAllCartSelectedRequest(isAllSelected));
+  };
+
+  const deleteCartItem = (id: number | string) => () => {
+    dispatch(deleteCartItemRequest(Number(id)));
+  };
+
+  const deleteAllCartItem = () => {
+    dispatch(deleteAllCartItemRequest());
+  };
+
+  return {
+    postCartItemQuantity,
+    updateCartItemQuantity,
+    selectCartItem,
+    selectAllCartItem,
+    deleteCartItem,
+    deleteAllCartItem,
+  };
+};
+
+export default useCartRequest;
```

Comment:
무슨 말씀이신지 이해했습니다.

- 관심사 분리를 위해 중간 메소드를 정의한 것에 불과하다. 
- 그러다보니 복잡도가 높아져 오히려 읽기가 힘들어졌다.
- 불필요하게 클로저까지 사용하여 더욱 복잡해진 것 같다

저도 다시 읽다보니, 중간과정 때문에 타고 들어가야하는 부분이 많아지고, 복잡도가 커진 것 같네요. 확실히 불필요하다고 생각해요..!
말씀하신 방향으로 수정해보았습니다.

[refactor: useCartRequest 훅 제거](https://github.com/woowacourse/react-shopping-cart/pull/119/commits/3045edf9b5cccb2fe4e195c06599ca484ad9a00b)

위 코멘트와 관련하여 질문드리고 싶은데요!

위 코멘트에서 로직을 "분리"한 것이 아니라 파일을 "나눈"것에 불과하다고 말씀해주셨는데요.
컨테이너-프리젠터 패턴이 비즈니스 로직과 UI를 다른 컴포넌트로 나누어서 코드를 작성하는 것처럼,
커스텀 훅으로 비슷하게 구현하고 싶었어요.
하지만 로이정님이 보기에는 이 방식이(useCartRequest 코드와 별개로..) 오히려 복잡도를 증가시키고, 읽기 어려워진다고 생각하시는 건지 궁금합니다.

그렇다면, 커스텀 훅은 로직을 추상화하여 재사용하는 데 의미가 있다고 생각하시는지 궁금합니다.
어떻게 사용해야할지 아직 감이 잘 안잡히네요 .. ㅠ

Reply:
Dan Abramov가 아래와 같이 이야기했었군요...

> Update from 2019: I wrote this article a long time ago and my views have since evolved. In particular, I don’t suggest splitting your components like this anymore. If you find it natural in your codebase, this pattern can be handy. But I’ve seen it enforced without any necessity and with almost dogmatic fervor far too many times. The main reason I found it useful was because it let me separate complex stateful logic from other aspects of the component. [Hooks](https://reactjs.org/docs/hooks-custom.html) let me do the same thing without an arbitrary division. This text is left intact for historical reasons but don’t take it too seriously.

비즈니스 로직 분리는 Hooks를 통해 가능해졌기 때문에, Container-Presenter을 구분지을 필요가 없다. 잘 이해했습니다!

훅이 재사용된다면, 의미가 충분히 있다고 볼 수 있고, 
재사용하지 않을 로직을 나누는 것에는 회의적인 입장이시군요. 

좋은 인사이트 주셔서 감사합니다!
더 고민해보겠습니다!!☺️